zeus21/sfrd.py

```diff
@@ -375,98 +599,283 @@ def __init__(self, Cosmo_Parameters, ClassCosmo, Astro_Parameters, HMF_interpola
+            duty_matom_component = np.exp(-Mh/Matom(z2))
+            fduty_III =  np.exp(-Mmol(Astro_Parameters, ClassCosmo, J21LW_interp, z2, vCB)/Mh) * duty_matom_component
+            
+        fstarM_III = fstarofz_III(Astro_Parameters, Cosmo_Parameters, z, Mh)
+        fstarM_III = np.fmin(fstarM_III, Astro_Parameters.fstarmax)
+        
+        return dMh_dt(Astro_Parameters, Cosmo_Parameters, HMF_interpolator, Mh, z) * fstarM_III * fduty_III
+    
+    
+def dMh_dt(Astro_Parameters, Cosmo_Parameters, HMF_interpolator, massVector, z):
+    # units of M_sun/yr
+    Mh = massVector
+    
+    if(Astro_Parameters.astromodel == 0): #GALLUMI-like
         if(Astro_Parameters.accretion_model == 0): #exponential accretion
-            dMhdz = Mh * constants.ALPHA_accretion_exponential
+            dMhdz = massVector * constants.ALPHA_accretion_exponential
+            
         elif(Astro_Parameters.accretion_model == 1): #EPS accretion
+            
             Mh2 = Mh * constants.EPSQ_accretion
-            indexMh2low = (Mh2 < Mh[0])
-            Mh2[indexMh2low] = Mh[0] #to avoid extrapolation to lower M. Those will have
-            sigmaMh = HMF_interpolator.sigma_int(Mh,z)
-            sigmaMh2 = HMF_interpolator.sigma_int(Mh2,z)
-            sigmaMh2[indexMh2low] = 1e99
-
+            indexMh2low = Mh2 < Mh.flatten()[0]
+            Mh2[indexMh2low] = Mh.flatten()[0]
+            
+            sigmaMh = HMF_interpolator.sigmaintlog((np.log(Mh), z))
+            sigmaMh2 = HMF_interpolator.sigmaintlog((np.log(Mh2), z))
+            sigmaMh2[np.full_like(sigmaMh2, fill_value=True, dtype = bool) * indexMh2low] = 1e99
+            
             growth = cosmology.growth(Cosmo_Parameters,z)
             dzgrow = z*0.01
             dgrowthdz = (cosmology.growth(Cosmo_Parameters,z+dzgrow) - cosmology.growth(Cosmo_Parameters,z-dzgrow))/(2.0 * dzgrow)
             dMhdz = - Mh * np.sqrt(2/np.pi)/np.sqrt(sigmaMh2**2 - sigmaMh**2) *dgrowthdz/growth * Cosmo_Parameters.delta_crit_ST
+            
         else:
             print("ERROR! Have to choose an accretion model in Astro_Parameters (accretion_model)")
-
-        Mhdot = dMhdz*cosmology.Hubinvyr(Cosmo_Parameters,z)*(1.0+z) #Msun/yr
-
-        return Mhdot * fstarM * fduty
+        Mhdot = dMhdz*cosmology.Hubinvyr(Cosmo_Parameters,z)*(1.0+z)
+        return Mhdot
 
     elif(Astro_Parameters.astromodel == 1): #21cmfast-like
-        fstarM = Astro_Parameters.fstar10*pow(Mh/1e10,Astro_Parameters.alphastar)
-        fstarM = Cosmo_Parameters.OmegaB/Cosmo_Parameters.OmegaM *np.fmin(fstarM,Astro_Parameters.fstarmax)
-        return Mh*fstarM/Astro_Parameters.tstar*cosmology.Hubinvyr(Cosmo_Parameters,z) * fduty
+        return Mh/Astro_Parameters.tstar*cosmology.Hubinvyr(Cosmo_Parameters,z)
     else:
         print('ERROR, MODEL is not defined')
+        
+
+def J_LW_Discrete(Astro_Parameters, Cosmo_Parameters, ClassCosmo, z, pop, rGreater, SFRD_II_interp, SFRD_III_cnvg_interp):
```

Comment:
This function is necessary for the gammaIII LW correction; it discretizes the integral into a sum over R instead of an interpolated integral over Z as used in the iteration process.